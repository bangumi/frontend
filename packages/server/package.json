{
  "name": "@bangumi/server",
  "version": "0.0.0",
  "private": true
}
