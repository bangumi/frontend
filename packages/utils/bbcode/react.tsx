import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
