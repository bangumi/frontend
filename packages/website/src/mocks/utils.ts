import { rest } from 'msw'
import fsp from 'fs/promises'
import path from 'path'

async function isFileExist (filePath: string): Promise<boolean> {
  try {
    await fsp.stat(filePath)
  } catch (e) {
    return false
  }

  return true
}

async function loadFixture (pathname: string, requestMethod: string): Promise<string> {
  const fixturePath = path.join(__dirname, './fixtures', `${pathname}-GET.json`)

  if (!(await isFileExist(fixturePath))) {
    const errMessage = `缺少对应 ${pathname} API Mock 文件。请创建 ${fixturePath} 添加 JSON mock`
    console.error(errMessage)
    throw new Error(errMessage)
  }

  return JSON.parse((await fsp.readFile(fixturePath)).toString())
}

type HTTPMethods = 'get' | 'post' | 'put' | 'delete' | 'options'

export function mockAPI (url: string, method: HTTPMethods) {
  return rest[method](url, async (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json(await loadFixture(req.url.pathname, req.method))
    )
  })
}
