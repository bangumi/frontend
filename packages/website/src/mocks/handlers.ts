import { mockAPI } from './utils'

const BASE_URL = 'https://api.bgm.tv'

function buildAPIURL (path: string) {
  return `${BASE_URL}${path}`
}

export const handlers = [
  mockAPI(buildAPIURL('/v0/subjects/:subjectId'), 'get'),
  mockAPI(buildAPIURL('/v0/persons/:personId/characters'), 'get'),
  mockAPI(buildAPIURL('/v0/characters/:characterId'), 'get')
]
