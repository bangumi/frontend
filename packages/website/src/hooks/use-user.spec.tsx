import { renderHook, waitFor } from '@testing-library/react';
import { rest } from 'msw';
import type { PropsWithChildren } from 'react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import { server as mockServer } from '../mocks/server';
import {
  CaptureError,
  LoginErrorCode,
  PasswordUnMatchError,
  UnknownError,
  UserProvider,
  useUser,
} from './use-user';

function mockLogin(
  statusCode: number,
  response: Object = {},
  headers: Record<string, string | string[]> = {},
): void {
  mockServer.use(
    rest.post('http://localhost:3000/p1/login', async (req, res, ctx) => {
      return res(ctx.status(statusCode), ctx.set(headers), ctx.json(response));
    }),
  );
}

const wrapper = ({ children }: PropsWithChildren) => (
  <MemoryRouter>
    <UserProvider>{children}</UserProvider>
  </MemoryRouter>
);

it.each([
  {
    statusCode: 401,
    body: { code: 'CAPTCHA_ERROR' },
    headers: { 'X-RateLimit-Remaining': '4' },
    expectedError: new CaptureError(4),
  },
  {
    statusCode: 401,
    body: { code: 'EMAIL_PASSWORD_ERROR' },
    headers: { 'X-RateLimit-Remaining': '4' },
    expectedError: new PasswordUnMatchError(4),
  },
  { statusCode: 400, body: { message: 'a' }, expectedError: new UnknownError('a') },
  { statusCode: 422, expectedError: new Error(LoginErrorCode.E_UNKNOWN_ERROR) },
  { statusCode: 418, expectedError: new Error(LoginErrorCode.E_UNKNOWN_ERROR) },
  { statusCode: 429, expectedError: new Error(LoginErrorCode.E_TOO_MANY_ERROR) },
  { statusCode: 502, expectedError: new Error(LoginErrorCode.E_UNKNOWN_ERROR) },
])(
  'should return error if request is failed with failed status $statusCode',
  async ({ statusCode, body = {}, headers = {}, expectedError }) => {
    const { result } = renderHook(() => useUser(), { wrapper });

    mockLogin(statusCode, body, headers);

    expect.assertions(1);
    await waitFor(async () => {
      await expect(result.current.login('fakeuser', 'fakepassword', 'fake-token')).rejects.toEqual(
        expectedError,
      );
    });
  },
);

it('should refresh me if login succeeded', async () => {
  const { result } = renderHook(() => useUser(), { wrapper });

  mockLogin(200);
  await waitFor(async () => {
    await result.current.login('fakeuser', 'fakepassword', 'fake-token');
    expect(result.current.user).toMatchSnapshot();
  });
});
