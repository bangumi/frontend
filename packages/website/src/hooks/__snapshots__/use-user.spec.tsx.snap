// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should refresh me if login succeeded 1`] = `
Object {
  "avatar": Object {
    "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
    "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
    "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
  },
  "id": 382951,
  "nickname": "树洞酱",
  "username": "382951",
}
`;
